### DISCUSSION TITLE
Memory leak when using @partial(jax.jit, static_argnums=(0,)) in class

### DISCUSSION THREAD

**Discussion by pipme** [jax-ml/jax#11855] [Q&A] [answered]
From the activity monitor, the process's memory keeps growing during the for loop. How to deal with it? `static_argnums` would cause memory leak? Thanks in advance.

```python
import jax
import jax.numpy as jnp
import numpy as np
from time import sleep
from functools import partial

class A:
    def __init__(self, name):
        self.name = name
        self.X = np.ones((5000, 5000))

    @partial(jax.jit, static_argnums=(0,))
    def f(self, Y):
        return jnp.sum(Y**2)

Y = np.ones(10)
for i in range(100):
    a = A(i)
    a.f(Y)
    sleep(2)
```

**Comment by jakevdp** [comment, marked as answer]
Hi - thanks for the question! In some senses this is intended behavior, in that the JAX jit cache uses strong references to track static arguments, so that if an argument goes out of scope the cache entry is not discarded. You can see why this would be useful, because, e.g. you wouldn't want to have to recompile a function just because you used a non-global string value as a static argument!

The side effect, then, is that if you have very large temporary static arguments, a strong reference will be created within the jit cache that will keep them in scope. This isn't really a leak per-se – the object is deliberately referenced in order to keep cached versions of compiled functions and prevent having to re-compile. You could address this by manually clearing the JIT cache of the function, in which case the temporary objects are garbage collected and the memory use remains constant:
```python
import os, psutil

for i in range(41):
  A.f.clear_cache()
  a = A(i)
  a.f(Y)
  if i % 10 == 0:
    process = psutil.Process(os.getpid())
    print(f"Iteration {i}: {process.memory_info().rss} bytes")
# Iteration 0: 471412736 bytes
# Iteration 10: 873562112 bytes
# Iteration 20: 873660416 bytes
# Iteration 30: 873660416 bytes
# Iteration 40: 873672704 bytes
```
The jit cache is an LRU cache of a fixed size, so eventually unused static arguments will be discarded, but if your static arguments are large, you'll probably run out of memory before that point.

A better solution here would be to use static argnums as JAX intends: i.e. pass small hashable truly static values. Note that your use of static_argnums here goes against the recommendations in the [`jax.jit` documentation](https://jax.readthedocs.io/en/latest/_autosummary/jax.jit.html); namely:

> `Static arguments should be hashable, meaning both __hash__ and __eq__ are implemented, and immutable.`

Your class does not implement `__hash__` or `__eq__`, and is not immutable (the value of `self.X` could change).

If you're working with classes that contain arrays, marking them as static is bad practice as it can easily result in surprises; here's a quick example:
```python
class A:
  def __init__(self):
    self.x = jnp.arange(5)

  @partial(jax.jit, static_argnums=0)
  def f(self):
    return self.x.sum()

a = A()
print(a.f())
# 10

a.x += 1 # mutation that doesn't change the hash
print(a.f())
# 10
```
The class has violated the contract of `static_argnums` – it is not immutable – and so the output of the second call is incorrect! It would be nice if JAX could catch this and warn users that it's happening, but that's a somewhat tricky thing to do well (see related discussion in #9024).

A better approach for this kind of thing would be to register your class as a [custom pytree](https://jax.readthedocs.io/en/latest/pytrees.html#extending-pytrees), which is effectively a way to tell jax which aspects of the class are static, and which are dynamic. For your class it might look like this:
```python
from jax.tree_util import register_pytree_node

class A:
  def __init__(self, name):
    self.name = name
    self.X = np.ones((5000, 5000))

  @jax.jit  # note: no static argnums
  def f(self, Y):
    return jnp.sum(Y**2)

def _flatten_A(a):
  children = (a.X,)  # non-static values
  aux_data = (a.name,)  # static values
  return (children, aux_data)

def _unflatten_A(aux_data, children):
  X, = children
  name, = aux_data
  a = A(name)
  a.X = X
  return A

register_pytree_node(A, _flatten_A, _unflatten_A)

a = A(1)
print(a.f(Y))
# 10.0
```
here `a.X` is no longer treated as a static argument, so no long-lived reference to it is created, and you'll no longer see the blow-up in memory on repeated calls.

Does that make sense?

**Reply by pipme** [reply]
Wow hi @jakevdp, thank you for the very detailed and helpful answer! I just tried with `A.f._clear_cache()` but it seems not working for me. I was running with jax 0.3.10 and just upgraded to 0.3.15. The output is:
```
Iteration 0: 362012672 bytes
Iteration 10: 2363199488 bytes     
Iteration 20: 4364181504 bytes
Iteration 30: 6365446144 bytes
Iteration 40: 8366460928 bytes
```
Thanks for the suggestion on custom pytree. It can definitely be a better approach and I will try to restructure my current stuff with that. For my case I have some Gaussian process parameters (in nested dictionary) to be optimized by jaxopt and some intermidiate cache values that depend on the input to `self.__init__` etc. for speeding up computations. I wasn't so familiar with the better practices by using pytree or frozen dataclass, therefore ended up using the class only as some namespace container and using some frozen dataclass here and there.  I would appreciate it if you have some reference project that follows a good jax style. I may need to have another look at `flax` or some other minimal and clean practices.

**Reply by pipme** [reply]
Oh `A.f._clear_cache()` doesn't work but `A.f.clear_cache()` does work. There was no `A.f.clear_cache()` in 0.3.10 so I thought it was a typo 😅.

**Reply by pipme** [reply]
I will have a closer look at your other comments tomorrow. Thanks again!

**Reply by pipme** [reply]
```python
from jax.tree_util import register_pytree_node


class A:
    def __init__(self, name):
        self.name = name
        self.X = np.ones((5000, 5000))

    @jax.jit  # note: no static argnums
    def f(self, Y):
        return jnp.sum(Y ** 2) * jnp.sum(self.X ** 2)


def _flatten_A(a):
    children = (a.X,)  # non-static values
    aux_data = (a.name,)  # static values
    return (children, aux_data)


def _unflatten_A(aux_data, children):
    (X,) = children
    (name,) = aux_data
    a = A(name)
    a.X = X
    return A

register_pytree_node(A, _flatten_A, _unflatten_A)

Y = jnp.ones(2)
a = A(1)
print(a.X.shape)  # output: (5000, 5000)
a.f(Y)  # AttributeError
```

I might still not fully understand. The above raises `AttributeError: type object 'A' has no attribute 'X'`.

**Reply by jakevdp** [reply]
Sorry, there's a typo. In `_unflatten_A`, it should be `return a` rather than `return A` (i.e. you return the unflattened object)

**Reply by pipme** [reply]
Oops get it. Thanks!

**Comment by patrick-kidger** [comment]
For context, cross-referencing https://github.com/patrick-kidger/diffrax/issues/142 here as another example of the same issue. (Where the solution is the same -- clear the JIT cache.)

In addition, @pipme - if you want to do "JAX with classes", and in particular have the register-as-custom-pytree handled for you, then have a look at [Equinox](https://github.com/patrick-kidger/equinox).

**Reply by pipme** [reply]
@patrick-kidger Thanks for the info! I was looking at [treeo](https://github.com/cgarciae/treeo) and will definitely have a look at [Equinox](https://github.com/patrick-kidger/equinox) too : )

**Comment by pipme** [comment]
```python
import jax
import jax.numpy as jnp
from jax.tree_util import register_pytree_node

class A:
    def __init__(self, name):
        self.name = name
        self.X = jnp.ones((5000, 5000))

    @jax.jit  # note: no static argnums
    def f(self, Y):
        return jnp.sum(Y ** 2) * jnp.sum(self.X ** 2)


def _flatten_A(a):
    children = (a.X,)  # non-static values
    aux_data = (a.name,)  # static values
    return (children, aux_data)


def _unflatten_A(aux_data, children):
    (X,) = children
    (name,) = aux_data
    a = A(name)
    a.X = X
    return a

register_pytree_node(A, _flatten_A, _unflatten_A)

a = A(1)
for i in range(1,5):
    a.f(jnp.ones(i))
print(A.f._cache_size())  # 4
del a
print(A.f._cache_size())  # 4
```
@jakevdp After registering the class as custom pytree, I would like the jit cache to be automatically cleared while it seems not. I am aware of https://github.com/google/jax/issues/7930#issuecomment-995003572. I need to create a lot of objects like `a` and call `a.f` on dynamic shape inputs, resulting non-negligible increasing memory due to jit cache. How could I deal with it?

Is calling `from xxx import A; A.f.clear_cache()` periodically the only way? I also had a look on #10828. Ideally we'd like to treat jitting class bound method in a different way and clear jit cache once the underlying `self` object is deleted. Am I missing something? Thanks for help!

**Reply by jakevdp** [reply]
Python methods are bound to classes, not to instances... you may be able to hack something along these lines by JIT-compiling a `lambda` function that's bound to the instance, and that goes out of scope when the instance is deleted

**Reply by jakevdp** [reply]
But to be honest, I'm not understanding why having a large compilation cache is problematic at this point. the contents of the `X` attribute are not being serialized, only its shape and dtype, so cache memory size shouldn't be an issue (it will max out at the LRU cache size limit of 4096 times a few bytes per entry). Am I missing something?

**Reply by pipme** [reply]
Thanks for clarification. In my case `A.f` involves some other heavier computations like `linalg.cholesky` and I would decorate both `jit` and `value_and_grad` for it. My program keeps growing memory and thus I suspect that it's due to jit cache. However I don't know how large in the memory the jit caches are. Is there a way to inspect/print it?

By setting `config.update("jax_disable_jit", True)`, the memory leak of my program is solved. So the cause should be JIT. But one mysterious thing is, when I mannually remove all `jit` from the code, without using `config.update("jax_disable_jit", True)`, the memory leak appears again! I get quite lost now. I am aware of some internal jax functions that by default use `jit`, e.g. `linalg._cholesky`, `linalg._solve_triangular`. I tried to call `clear_cache` for them periodically but the memory leak still exists.

For comparison, I also have a look at  `jax.lib.xla_bridge.get_backend().live_buffers()` for both 
1. `config.update("jax_disable_jit", True)`. 
2. Manually remove jit in my code and call `clear_cache` for `linalg._cholesky` etc. 

The lengths and shapes of live buffers are the same and both grow during execution, since I have some bookkeeping for the results of each iteration. Live buffers only represent the created `DeviceArray` if I understand correctly.

Is there anything you suggest to further look into for solving this memory issue?  Thanks! @jakevdp

(Btw, I use [treeo](https://github.com/cgarciae/treeo) to register custom tree and there are also other functions not related to `A` for jitting and taking `value_and_grad`, I have functions like 
```python
class B:
    def __init__(self):
        self.large_constant = np.ones((100,100))  # a large array
    
    def build_f(self):
        large_constant = self.constant
        def fn(x):
            return jnp.sum(large_constant) + jnp.sum(x)  # some heavy computations using both constant and x
        return fn

b = B()
f = jax.jit(jax.value_and_grad(b.build_f()))
``` 
Is writing in the above way also not a good practice since the large constant is baked(?) to compilation? But I still don't understand why case 2. has memory leak.

**Reply by jakevdp** [reply]
It's hard to guess why you might have a memory leak without seeing the code. Can you create a [self-contained minimal example](https://stackoverflow.com/help/minimal-reproducible-example) that demonstrates the issue?

**Reply by pipme** [reply]
Thanks. It's a bit complex to create a minimal example for my concrete problem. In the mean time, a simple snippet below increases memory by **~20MB** at each print, until 4096 maxmium cache size. It starts with 112MB and finalize with 637MB, which is quite big and non-negligible as far as I can see. As in my problem, the array shape keeps changing. Is there a way to clear all cache programmatically? I saw https://github.com/google/jax/issues/2072#issuecomment-578421838 but it looks like `xla._xla_callable.cache_clear()` has been removed. Or a method to config lru's `maxsize` can also be useful?

```python
@jax.jit
@jax.value_and_grad
def f(x):
    return jnp.sum(x)


print(f"{psutil.Process().memory_info().rss / 1e6} MB")
for i in tqdm(range(5000)):
    f(jnp.ones(i + 1))
    if i % 100 == 0:
        print(f"{psutil.Process().memory_info().rss / 1e6} MB")
print(f"{psutil.Process().memory_info().rss / 1e6} MB")
```

**Reply by pipme** [reply]
Now I am trying with a hack https://github.com/google/jax/issues/10828#issuecomment-1138428231 provided by @patrick-kidger . It reduces the memory usage quite a lot during the long run!

**Reply by pipme** [reply]
Ok I see `clear_backend()`. Will have a try.
https://github.com/google/jax/blob/7721579700ee5c4a951e72156a0bdac4c9768f43/jax/_src/api.py#L3250-L3264

**Reply by jakevdp** [reply]
Thanks for the example - it looks like each cached compilation is using ~200KB of memory. If this is problematic, then the new `clear_backends` API introduced in https://github.com/google/jax/pull/11462 may be helpful...

At this point the cache size is fixed at 4096 (I believe) but perhaps this makes clear that we should allow this to be user-configurable?
